### DISCUSSION TITLE
Misused header name, 'Content-Type'. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects.'

### DISCUSSION THREAD

**Discussion by kasunf** [gabrieldwight/Whatsapp-Business-Cloud-Api-Net#38] [Q&A] [answered]
I got this error when sending filedata    

       Misused header name, 'Content-Type'. Make sure request headers are used with HttpRequestMessage, response headers with 
       HttpResponseMessage, and content headers with HttpContent objects.'

Here is my code
        var httpClient = new HttpClient();
    
        httpClient.BaseAddress = WhatsAppBusinessRequestEndpoint.BaseAddress;
        
    
        var whatsAppBusinessClient = new WhatsAppBusinessClient(httpClient, whatsAppConfig); 

        var media = await whatsAppBusinessClient.UploadFileDataAsync("1",path, "application/pdf");

        var url = await whatsAppBusinessClient.GetMediaUrlAsync(media.Id);

        DocumentMessageByUrlRequest documentMessage = new DocumentMessageByUrlRequest();
        documentMessage.To =No;
        documentMessage.Document = new MediaDocumentUrl();
        documentMessage.Document.Link = url.Url;

        var results = await whatsAppBusinessClient.SendDocumentAttachmentMessageByUrlAsync(documentMessage);

**Comment by gabrieldwight** [comment]
Hi @kasunf,  The function above **UploadFileDataAsync** was for the resumable upload API, you will need to create an upload session Id first. I have tested it with image files only. The same steps should work with document files once you have created an upload session id.

```C#
var uploadSessionId = await whatsAppBusinessClient.CreateResumableUploadSessionAsync(long fileLength, string fileType, string fileName);

var media = await whatsAppBusinessClient.UploadFileDataAsync(uploadSessionId, path, "application/pdf");

var url = await whatsAppBusinessClient.GetMediaUrlAsync(media.Id);
```

Alternatively, you can use the uploadMediaFunction
```C#
UploadMediaRequest uploadMediaRequest = new UploadMediaRequest();
uploadMediaRequest.File = path;
uploadMediaRequest.Type = "application/pdf";

var uploadMediaResult = await _whatsAppBusinessClient.UploadMediaAsync(uploadMediaRequest);

var url = await whatsAppBusinessClient.GetMediaUrlAsync(media.Id);
```

**Reply by kasunf** [reply]
What is fileType for PDF here?

**Reply by gabrieldwight** [reply]
Content type: application/pdf

**Reply by kasunf** [reply]
Line
    var media = await whatsAppBusinessClient.UploadFileDataAsync(uploadSessionId, path, "application/pdf")
caused 

Misused header name, 'Content-Type'. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects

**Reply by kasunf** [reply]
tried using uploadMediaFunction, no error occurred but whatsapp msg not received to the test no.

**Reply by kasunf** [reply]
Code Snippet , error line 7
      
        var httpClient = new HttpClient();
        httpClient.BaseAddress = WhatsAppBusinessRequestEndpoint.BaseAddress;
        string fileName = path;
        FileInfo fi = new FileInfo(fileName);
        var whatsAppBusinessClient = new WhatsAppBusinessClient(httpClient, whatsAppConfig); 
        var uploadSessionId = await whatsAppBusinessClient.CreateResumableUploadSessionAsync(fi.Length, "application/pdf", fi.Name );
        var media = await whatsAppBusinessClient.UploadFileDataAsync(uploadSessionId.Id, path,"application/pdf");

        var url = await whatsAppBusinessClient.GetMediaUrlAsync(media.Id);

        
        DocumentMessageByUrlRequest documentMessage = new DocumentMessageByUrlRequest();
        documentMessage.To =<No>;
        documentMessage.Document = new MediaDocumentUrl();
        documentMessage.Document.Link = url.Url;

        var results = await whatsAppBusinessClient.SendDocumentAttachmentMessageByUrlAsync(documentMessage);

**Reply by gabrieldwight** [reply]
I will have a look at it to see if there was a change on WhatsApp end for resumable upload api and issue a fix. 

Did the uploadmedia return a media id for the document?

**Reply by kasunf** [reply]
yes.

**Reply by kasunf** [reply]
> yes.

wamid.HBgLOTQ3MDYyNDU5ODEVAgARGBI0NTA1ODVEMDMyQTdBQ0E2MUIA  received this response to the SendDocumentAttachmentMessageByUrlAsync

**Reply by kasunf** [reply]
> > yes.
> 
> wamid.HBgLOTQ3MDYyNDU5ODEVAgARGBI0NTA1ODVEMDMyQTdBQ0E2MUIA received this response to the SendDocumentAttachmentMessageByUrlAsync

@gabrieldwight 

This is uploadmediafunction code snippet

        var httpClient = new HttpClient();

        httpClient.BaseAddress = WhatsAppBusinessRequestEndpoint.BaseAddress;

        var whatsAppBusinessClient = new WhatsAppBusinessClient(httpClient, whatsAppConfig); 

        UploadMediaRequest uploadrequest = new UploadMediaRequest();
        uploadrequest.File = path;
        uploadrequest.Type = "application/pdf";
        var media = await whatsAppBusinessClient.UploadMediaAsync(uploadrequest);

        var url = await whatsAppBusinessClient.GetMediaUrlAsync(media.MediaId);


        DocumentMessageByUrlRequest documentMessage = new DocumentMessageByUrlRequest();
        documentMessage.To = No;
        documentMessage.Document = new MediaDocumentUrl();
        documentMessage.Document.Link = url.Url;

        var results = await whatsAppBusinessClient.SendDocumentAttachmentMessageByUrlAsync(documentMessage);

**Reply by gabrieldwight** [reply]
That's looks fine.

Apart from document file, are you able to receive an image message using media id on the test phone number?

**Reply by kasunf** [reply]
> That's looks fine.
> 
> Apart from document file, are you able to receive an image message using media id on the test phone number?

same results. no error but msg not delivered.

**Reply by kasunf** [reply]
> That's looks fine.
> 
> Apart from document file, are you able to receive an image message using media id on the test phone number?

i got only the text messages. No media , No document messages

**Reply by kasunf** [reply]
> > That's looks fine.
> > Apart from document file, are you able to receive an image message using media id on the test phone number?
> 
> i got only the text messages. No media , No document messages

I am using MAUI .net for android.

**Reply by gabrieldwight** [reply]
I have pushed the fix for the content-type error, The function that you were using to upload media using session id will return an encoded header file that is mostly used to create template media messages with the encoded header response since it won't contain the media id.

Resumable upload returns this type of response
```JSON
{"h":"2:c2FtcGxlLm1wNA==:image/jpeg:GKAj0gAUCZmJ1voFADip2iIAAAAAbugbAAAA:e:1472075513:ARZ_3ybzrQqEaluMUdI"}
``` 

Sending document message with media Id 
Code snippet
```C#
var httpClient = new HttpClient();

httpClient.BaseAddress = WhatsAppBusinessRequestEndpoint.BaseAddress;

    var whatsAppBusinessClient = new WhatsAppBusinessClient(httpClient, whatsAppConfig); 

    UploadMediaRequest uploadrequest = new UploadMediaRequest();
    uploadrequest.File = path;
    uploadrequest.Type = "application/pdf";
    var media = await whatsAppBusinessClient.UploadMediaAsync(uploadrequest);

DocumentMessageByIdRequest documentMessage = new DocumentMessageByIdRequest();
                            documentMessage.To = No;
                            documentMessage.Document = new MediaDocument();
                            documentMessage.Document.Id = media.MediaId;
                            documentMessage.Document.Caption = caption;

                            results = await _whatsAppBusinessClient.SendDocumentAttachmentMessageByIdAsync(documentMessage);
```

Note that the retrieving media URL from the media Id is usually valid for 5 minutes. Alternatively, if you have a URL that hosts your document you can use send message using your document endpoint. You can try to see if it works .NET MAUI for android.

**Reply by kasunf** [reply]
Thank you for your effort, I Updated to new version, but this didn't work for me.

**Reply by gabrieldwight** [reply]
Let me try to prepare a .NET MAUI sample to check how it handles media uploads then I will let you know

**Reply by kasunf** [reply]
here is a MAUI sample project   https://github.com/kasunf/WhatsApp-MAUI , Can you please try it?

**Reply by gabrieldwight** [reply]
Let me try and test it then I get back to you

**Reply by gabrieldwight** [reply, marked as answer]
On my side, it is working with the MAUI you sent using the file picker API
```C#
var customFileType = new FilePickerFileType(
			new Dictionary<DevicePlatform, IEnumerable<string>>
				{
					{ DevicePlatform.iOS, new[] { "" } }, // UTType values
                    { DevicePlatform.Android, new[] { "image/jpeg", "image/png", "application/pdf" } }, // MIME type
                    { DevicePlatform.WinUI, new[] { ".png", ".jpg", ".pdf" } }, // file extension
                    { DevicePlatform.Tizen, new[] { "*/*" } },
					{ DevicePlatform.macOS, new[] { "" } }, // UTType values
                });

		PickOptions options = new()
		{
			PickerTitle = "Please select a media file",
			FileTypes = customFileType,
		};

		FileResult fileResult = await PickAndShow(options);

		string path = fileResult.FullPath;

string fileName = path;
        FileInfo fi = new FileInfo(fileName);
        UploadMediaRequest uploadrequest = new UploadMediaRequest();
        uploadrequest.File = path;
		uploadrequest.Type =  fileResult.ContentType; //"application/pdf";
        var media =  await whatsAppBusinessClient.UploadMediaAsync(uploadrequest);

if (fi.Extension.Contains("jpg") || fi.Extension.Contains("png"))
		{
			ImageMessageByIdRequest imageMessage = new ImageMessageByIdRequest();
			imageMessage.To = no;
			imageMessage.Image = new MediaImage();
			imageMessage.Image.Id = media.MediaId;
			imageMessage.Image.Caption = "MAUI Image Test";

			await whatsAppBusinessClient.SendImageAttachmentMessageByIdAsync(imageMessage);
		}
		else
		{
			DocumentMessageByIdRequest documentMessage = new DocumentMessageByIdRequest();
			documentMessage.To = no;
			documentMessage.Document = new MediaDocument();
			documentMessage.Document.Id = media.MediaId;
			documentMessage.Document.Caption = "MAUI PDF Test"; //"hello";
			documentMessage.Document.Filename = fi.Name; //"hello.pdf";


			await whatsAppBusinessClient.SendDocumentAttachmentMessageByIdAsync(documentMessage);
		}
```
File Picker API
```C#
public async Task<FileResult> PickAndShow(PickOptions options)
	{
		try
		{
			var result = await FilePicker.Default.PickAsync(options);
			/*if (result != null)
			{
				if (result.FileName.EndsWith("jpg", StringComparison.OrdinalIgnoreCase) ||
					result.FileName.EndsWith("png", StringComparison.OrdinalIgnoreCase))
				{
					using var stream = await result.OpenReadAsync();
					var image = ImageSource.FromStream(() => stream);
				}
				else
				{
					using var stream = await result.OpenReadAsync();
				}
			}*/

			return result;
		}
		catch (Exception ex)
		{
			// The user canceled or something went wrong
		}

		return null;
	}
```

Results:
Media Message
![image](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/28096255/2e1a6d0f-5488-44e3-b9b2-849046e4e040)


Document Message
![image](https://github.com/gabrieldwight/Whatsapp-Business-Cloud-Api-Net/assets/28096255/5b118b2d-9149-444c-82b0-1e136a75841b)

**Reply by kasunf** [reply]
This is working perfectly. I changed the path to auto generated invoice and it worked. Thank you so much.
